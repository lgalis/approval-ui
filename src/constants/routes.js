const routes = {
  allrequests: {
    index: '/allrequests'
  },
  requests: {
    index: '/requests',
    addComment: '/requests/add-comment',
    approve: '/requests/approve',
    deny: '/requests/deny'
  },
  request: {
    index: '/request',
    addComment: '/request/add-comment',
    approve: '/request/approve',
    deny: '/request/deny'
  },
  allrequest: {
    index: '/allrequest',
    addComment: '/allrequest/add-comment',
    approve: '/allrequest/approve',
    deny: '/allrequest/deny'
  },
  workflows: {
    index: '/workflows',
    add: '/workflows/add-workflow',
    remove: '/workflows/remove',
    edit: '/workflows/edit'
  }
};

export default routes;
