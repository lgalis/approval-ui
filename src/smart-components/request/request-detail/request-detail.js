import React, { Fragment, useContext, useEffect, useReducer } from 'react';
import { Route, useLocation, Switch } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { Grid, GridItem } from '@patternfly/react-core';
import { Section } from '@redhat-cloud-services/frontend-components/components/cjs/Section';
import ActionModal from '../action-modal';
import RequestInfoBar from './request-info-bar';
import RequestTranscript from './request-transcript';
import { fetchRequest, fetchRequestContent } from '../../../redux/actions/request-actions';
import { RequestLoader } from '../../../presentational-components/shared/loader-placeholders';
import { TopToolbar, TopToolbarTitle } from '../../../presentational-components/shared/top-toolbar';
import UserContext from '../../../user-context';
import useQuery from '../../../utilities/use-query';
import { approvalPersona } from '../../../helpers/shared/helpers';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import requestsMessages from '../../../messages/requests.messages';

const initialState = {
  isFetching: true
};

const requestState = (state, action) => {
  switch (action.type) {
    case 'setFetching':
      return { ...state, isFetching: action.payload };
    default:
      return state;
  }
};

const RequestDetail = ({ requestBreadcrumbs, indexpath }) => {
  const [{ isFetching }, stateDispatch ] = useReducer(requestState, initialState);

  const { selectedRequest, requestContent } = useSelector(
    ({
      requestReducer: {
        requestContent: requestContent,
        selectedRequest: selectedRequest
      }
    }) => ({ selectedRequest, requestContent })
  );

  const [{ request: id }] = useQuery([ 'request' ]);
  const location = useLocation();
  const dispatch = useDispatch();
  const { userRoles: userRoles } = useContext(UserContext);
  const intl = useIntl();

  useEffect(() => {
    Promise.all([ dispatch(fetchRequest(id, approvalPersona(userRoles))), dispatch(fetchRequestContent(id, approvalPersona(userRoles))) ])
    .then(() => stateDispatch({ type: 'setFetching', payload: false }));
  }, []);

  const updateRequest = (id) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return dispatch(fetchRequest(id, approvalPersona(userRoles))), dispatch(fetchRequestContent(id, approvalPersona(userRoles)))
    .then(() => stateDispatch({ type: 'setFetching', payload: false }))
    .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  const renderRequestDetails = () => {
    if (isFetching || !selectedRequest || Object.keys(selectedRequest).length === 0) {
      return (
        <Section style={ { backgroundColor: 'white', minHeight: '100%' } }>
          <RequestLoader />
        </Section>
      );
    }
    else {
      return (
        <Fragment>
          <GridItem md={ 4 } lg={ 3 } className="info-bar pf-u-p-0">
            <RequestInfoBar request={ selectedRequest } requestContent={ requestContent }/>
          </GridItem>
          <GridItem md={ 8 } lg={ 9 } className="detail-pane pf-u-p-lg">
            <RequestTranscript request={ selectedRequest } url={ location.url } indexpath={ indexpath }/>
          </GridItem>
        </Fragment>
      );
    }
  };

  return (
    <Fragment>
      <Switch>
        <Route exact path={ indexpath.addComment }>
          <ActionModal actionType={ 'Add Comment' }
            postMethod={ () => updateRequest(selectedRequest.id) }
            closeUrl={ { pathname: indexpath.index, search: `?request=${selectedRequest.id}` } }/>
        </Route>
        <Route exact path={ indexpath.approve } render={ props =>
          <ActionModal { ...props } actionType={ 'Approve' }
            postMethod={ () => updateRequest(selectedRequest.id) }
            closeUrl={ { pathname: indexpath.index, search: `?request=${selectedRequest.id}` } } /> } />
        <Route exact path={ indexpath.deny } render={ props =>
          <ActionModal { ...props } actionType={ 'Deny' }
            postMethod={ () => updateRequest(selectedRequest.id) }
            closeUrl={ { pathname: indexpath.index, search: `?request=${selectedRequest.id}` } } /> } />
      </Switch>
      <TopToolbar
        breadcrumbs={ requestBreadcrumbs }
        paddingBottom={ true }
      >
        <TopToolbarTitle title={ intl.formatMessage(requestsMessages.requestTitle, { id }) } />
      </TopToolbar>
      <Section type="content">
        <Grid hasGutter>
          { renderRequestDetails() }
        </Grid>
      </Section>
    </Fragment>
  );
};

RequestDetail.propTypes = {
  requestBreadcrumbs: PropTypes.array,
  indexpath: PropTypes.object
};

export default RequestDetail;
