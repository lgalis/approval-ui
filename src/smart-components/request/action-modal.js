import React from 'react';
import PropTypes from 'prop-types';
import FormRenderer from '../common/form-renderer';
import { useHistory } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal } from '@patternfly/react-core';
import { createRequestAction } from '../../redux/actions/request-actions';
import { createRequestCommentSchema } from '../../forms/request-comment-form.schema';
import useQuery from '../../utilities/use-query';
import routes from '../../constants/routes';
import { useIntl } from 'react-intl';
import actionModalMessages from '../../messages/action-modal.messages';
import requestsMessages from '../../messages/requests.messages';

const actionTypeToMsg = {
  Approve: requestsMessages.approveTitle,
  Deny: requestsMessages.denyTitle
};

const ActionModal = ({
  actionType,
  createRequestAction,
  closeUrl,
  postMethod
}) => {
  const intl = useIntl();
  const { push } = useHistory();
  const [{ request: id }] = useQuery([ 'request' ]);
  const onSubmit = (data) => {
    const operationType = { 'Add Comment': 'memo', Approve: 'approve', Deny: 'deny' };
    const actionName = actionType === 'Add Comment'
      ? intl.formatMessage(requestsMessages.addCommentTitle)
      : intl.formatMessage(actionModalMessages.actionName, { actionType: intl.formatMessage(actionTypeToMsg[actionType]) }) ;

    return postMethod ?
      createRequestAction(
        actionName,
        id,
        { operation: operationType[actionType], ...data },
        intl
      ).then(() => postMethod()).then(() => push(closeUrl))
      : createRequestAction(
        actionName,
        id,
        { operation: operationType[actionType], ...data },
        intl
      ).then(() => push(closeUrl));
  };

  const onCancel = () => push(closeUrl);

  return (
    <Modal
      variant="large"
      title={ actionType === 'Add Comment'
        ? intl.formatMessage(actionModalMessages.requestTitle, { id })
        : intl.formatMessage(actionModalMessages.requestActionTitle, { id, actionType: intl.formatMessage(actionTypeToMsg[actionType]) })
      }
      isOpen
      onClose={ onCancel }
    >
      <FormRenderer
        schema={ createRequestCommentSchema(actionType === 'Deny', intl) }
        onSubmit={ onSubmit }
        onCancel={ onCancel }
      />
    </Modal>
  );
};

ActionModal.defaultProps = {
  closeUrl: routes.requests.index
};

ActionModal.propTypes = {
  createRequestAction: PropTypes.func.isRequired,
  postMethod: PropTypes.func,
  actionType: PropTypes.string,
  closeUrl: PropTypes.oneOfType([ PropTypes.string, PropTypes.shape({ patname: PropTypes.string, search: PropTypes.string }) ])
};

const mapDispatchToProps = (dispatch) => bindActionCreators({
  createRequestAction
}, dispatch);

export default connect(null, mapDispatchToProps)(ActionModal);
