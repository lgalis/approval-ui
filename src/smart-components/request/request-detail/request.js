import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { useIsApprovalAdmin } from '../../../helpers/shared/helpers';
import { ActionTranscript } from './action-transcript';

import {
  Stack,
  StackItem,
  DataListItem,
  DataListItemRow,
  DataListCell,
  DataListToggle,
  DataListItemCells,
  DataListContent,
  TextVariants,
  TextContent
} from '@patternfly/react-core';
import UserContext from '../../../user-context';
import routes from '../../../constants/routes';
import { useIntl } from 'react-intl';
import requestsMessages from '../../../messages/requests.messages';
import { untranslatedMessage } from '../../../utilities/constants';
import RequestActions from '../request-actions';

export const Request = ({ item, isExpanded, toggleExpand, indexpath }) => {
  const { userRoles: userRoles } = useContext(UserContext);
  const isApprovalAdmin = useIsApprovalAdmin(userRoles);
  const intl = useIntl();

  const checkCapability = (item, capability) => {
    if (isApprovalAdmin) {
      return true;
    }

    return item.metadata && item.metadata.user_capabilities && item.metadata.user_capabilities[capability];
  };

  return (
    <DataListItem key={ `request-${item.id}` }
      aria-labelledby={ `check-request-${item.id}` }
      isExpanded={ isExpanded }>
      <DataListItemRow>
        <DataListToggle
          onClick={ () => toggleExpand(`request-${item.id}`) }
          isExpanded={ isExpanded }
          id={ `request-${item.id}` }
          aria-labelledby={ `request-${item.id} request-${item.id}` }
          aria-label={ intl.formatMessage(requestsMessages.toogleDetailsFor) }
        />
        <DataListItemCells
          dataListCells={ [
            <DataListCell key={ item.id }>
              <span id={ `${item.id}-name` }>{ item.group_name ? item.group_name : item.name }</span>
            </DataListCell>,
            <DataListCell key={ `${item.id}-state` }>
              <span style={ { textTransform: 'capitalize' } } id={ `${item.id}-state` }>
                { intl.formatMessage(requestsMessages[item.state] || untranslatedMessage(item.state)) }
              </span>
            </DataListCell>,
            <DataListCell key={ `${item.id}-action` }>
              <RequestActions
                approveLink={ indexpath.approve }
                denyLink={ indexpath.deny }
                commentLink={ indexpath.addComment }
                request={ item }
                canApproveDeny={ checkCapability(item, 'approve') }
                canComment={ checkCapability(item, 'memo') }
              />
            </DataListCell>
          ] }/>
      </DataListItemRow>
      <DataListContent aria-label={ intl.formatMessage(requestsMessages.requestContentDetails) }
        isHidden={ !isExpanded }>
        <Stack hasGutter>
          <StackItem>
            <TextContent component={ TextVariants.h6 }>
              <ActionTranscript actionList={ item.actions }/>
            </TextContent>
          </StackItem>
        </Stack>
      </DataListContent>
    </DataListItem>
  );
};

Request.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    state: PropTypes.string,
    actions: PropTypes.array,
    group_name: PropTypes.string.isRequired,
    requestActions: PropTypes.shape({
      data: PropTypes.array
    }),
    metadata: PropTypes.shape({
      user_capabilities: PropTypes.object
    })
  }).isRequired,
  isExpanded: PropTypes.bool.isRequired,
  toggleExpand: PropTypes.func.isRequired,
  indexpath: PropTypes.object
};

Request.defaultProps = {
  indexpath: routes.request
};
