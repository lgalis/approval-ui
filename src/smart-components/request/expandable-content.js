import React, { Fragment, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { TextContent, Text, TextVariants, Level, LevelItem, Button, Bullseye } from '@patternfly/react-core';
import { isRequestStateActive } from '../../helpers/shared/helpers';
import { fetchRequestContent } from '../../helpers/request/request-helper';
import { Spinner } from '@redhat-cloud-services/frontend-components';

export const ExpandedItem = ({ title = '', detail = '' }) => (
  <TextContent>
    <Text className="data-table-detail heading" component={ TextVariants.small }>{ title }</Text>
    <Text className="data-table-detail content" component={ TextVariants.h5 }>{ detail }</Text>
  </TextContent>
);

ExpandedItem.propTypes = {
  title: PropTypes.node,
  detail: PropTypes.node
};

const ExpandableContent = ({ id, number_of_children, state, reason }) => {
  const requestActive = isRequestStateActive(state) && number_of_children === 0;
  const [ requestContent, setRequestContent ] = useState([]);
  const [ isLoading, setIsLoading ] = useState(true);
  const [ fetchStarted, setIsFetching ] = useState(false);

  const expandedRequests = useSelector(({ requestReducer: { expandedRequests }}) => expandedRequests);

  useEffect(() => {
    if (!fetchStarted && isLoading && expandedRequests.includes(id)) {
      setIsFetching(true);
      fetchRequestContent(id).then((data) => { setRequestContent(data); setIsLoading(false); }).catch(() => setIsLoading(false));
    }
  }, [ expandedRequests ]);

  if (isLoading) {
    return (<Bullseye>
      <Spinner/>
    </Bullseye>);
  }

  return (
    <Fragment>
      <Level>
        <LevelItem>
          <ExpandedItem title="Product" detail={ requestContent ? requestContent.product : 'Unknown' } />
        </LevelItem>
        { requestActive && <LevelItem>
          <Link to={ `/requests/approve/${id}` }  className="pf-u-mr-md">
            <Button variant="primary" aria-label="Approve Request" isDisabled={ !requestActive }>
              Approve
            </Button>
          </Link>
          <Link to={ `/requests/deny/${id}` }>
            <Button variant="danger" aria-label="Deny Request">
              Deny
            </Button>
          </Link>
        </LevelItem>
        }</Level>
      <Level>
        <LevelItem>
          <ExpandedItem title="Portfolio" detail={ requestContent ? requestContent.portfolio : 'Unknown' }/>
          <ExpandedItem title="Platform" detail={ requestContent ? requestContent.platform : 'Unknown' }/>
          <ExpandedItem title="Reason" detail={ reason ? reason : '' }/>
        </LevelItem>
      </Level>
    </Fragment>
  );
};

ExpandableContent.propTypes = {
  id: PropTypes.string,
  content: PropTypes.object,
  number_of_children: PropTypes.number,
  uname: PropTypes.string,
  state: PropTypes.string,
  reason: PropTypes.string
};

export default ExpandableContent;
