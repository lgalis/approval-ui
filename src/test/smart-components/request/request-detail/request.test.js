import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import { Request } from '../../../../smart-components/request/request-detail/request';
import routes from '../../../../constants/routes';
import { IntlProvider } from 'react-intl';

const ComponentWrapper = ({ children }) => (
  <IntlProvider locale="en">
    <MemoryRouter initialEntries={ [ '/foo' ] }>
      { children }
    </MemoryRouter>
  </IntlProvider>
);

describe('<Request />', () => {
  let initialProps;
  beforeEach(() => {
    initialProps = {
      isExpanded: false,
      toggleExpand: jest.fn(),
      item: {
        id: '111',
        group_name: 'Group',
        state: 'notified',
        metadata: {
          user_capabilities: { approve: true, deny: true }
        },
        actions: [
          {
            id: '1',
            operation: 'start',
            comments: null,
            created_at: '2020-01-29T17:08:56.850Z',
            processed_by: 'system'
          },
          {
            id: '2',
            operation: 'notify',
            comments: null,
            created_at: '2020-01-29T17:09:14.994Z',
            processed_by: 'system'
          }
        ]
      }
    };
  });

  it('should render correctly', () => {
    const initialPropsNoDate = {
      isExpanded: false,
      toggleExpand: jest.fn(),
      item: {
        id: '111',
        state: 'no-state',
        group_name: 'Group1',
        metadata: {
          user_capabilities: { approve: true, deny: true }
        },
        actions: [
          {
            id: '1',
            operation: 'start',
            comments: null,
            processed_by: 'system'
          },
          {
            id: '2',
            operation: 'notify',
            comments: null,
            processed_by: 'system'
          }
        ]
      }
    };
    const wrapper = mount(<ComponentWrapper><Request { ...initialPropsNoDate } /></ComponentWrapper>).find(Request);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render item in active state with link', () => {
    const wrapper = mount(
      <ComponentWrapper>
        <Request
          { ...initialProps }
          item={ {
            id: '111',
            state: 'notified',
            group_name: 'Group1',
            metadata: {
              user_capabilities: { approve: true, deny: true }
            },
            actions: [
              {
                id: '1',
                operation: 'start',
                comments: null,
                created_at: '2020-01-29T17:08:56.850Z',
                processed_by: 'system'
              },
              {
                id: '2',
                operation: 'notify',
                comments: null,
                created_at: '2020-01-29T17:09:14.994Z',
                processed_by: 'system'
              }
            ]} }
        />
      </ComponentWrapper>
    );
    wrapper.update();
    expect(wrapper.find('a#approve-111')).toHaveLength(1);
    expect(wrapper.find('a#deny-111')).toHaveLength(1);

    wrapper.find('Link#approve-111').simulate('click', { button: 0 });
    wrapper.update();
    const history = wrapper.find(MemoryRouter).instance().history;
    expect(history.location.pathname).toEqual(routes.request.approve);
    expect(history.location.search).toEqual('?request=111');
    wrapper.find('Link#deny-111').simulate('click', { button: 0 });
    expect(history.location.pathname).toEqual(routes.request.deny);
    expect(history.location.search).toEqual('?request=111');
  });

  it('should expand item', () => {
    const toggleExpand = jest.fn();
    const wrapper = mount(
      <ComponentWrapper>
        <Request
          { ...initialProps }
          item={ {
            id: '111',
            state: 'notified',
            group_name: 'Group1',
            actions: []
          } }
          isActive
          toggleExpand={ toggleExpand }
        />
      </ComponentWrapper>
    );
    wrapper.update();
    wrapper.find('button.pf-c-button.pf-m-plain').simulate('click');
    wrapper.update();
    expect(toggleExpand).toHaveBeenCalledWith('request-111');
  });

  it('should open comment modal', () => {
    const wrapper = mount(
      <ComponentWrapper>
        <Request
          { ...initialProps }
          item={ {
            id: '111',
            state: 'notified',
            group_name: 'Group1',
            metadata: {
              user_capabilities: { approve: true, deny: true, memo: true }
            },
            actions: []
          } }
          isActive
        />
      </ComponentWrapper>
    );
    wrapper.update();

    wrapper.find('a#comment-111').first().simulate('click', { button: 0 });
    wrapper.update();

    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual(routes.request.addComment);
    expect(wrapper.find(MemoryRouter).instance().history.location.search).toEqual('?request=111');
  });

  it('should open approve modal', () => {
    const wrapper = mount(
      <ComponentWrapper>
        <Request
          { ...initialProps }
          item={ {
            id: '111',
            state: 'notified',
            group_name: 'Group1',
            metadata: {
              user_capabilities: { approve: true, deny: true, memo: true }
            },
            actions: []
          } }
          isActive
        />
      </ComponentWrapper>
    );
    wrapper.update();
    wrapper.find('a#approve-111').first().simulate('click', { button: 0 });
    wrapper.update();

    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual(routes.request.approve);
    expect(wrapper.find(MemoryRouter).instance().history.location.search).toEqual('?request=111');
  });

  it('should open deny modal', () => {
    const wrapper = mount(
      <ComponentWrapper>
        <Request
          { ...initialProps }
          item={ {
            id: '111',
            state: 'notified',
            group_name: 'Group1',
            metadata: {
              user_capabilities: { approve: true, deny: true, memo: true }
            },
            actions: []
          } }
          isActive
        />
      </ComponentWrapper>
    );
    wrapper.update();
    wrapper.find('a#deny-111').first().simulate('click', { button: 0 });
    wrapper.update();

    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual(routes.request.deny);
    expect(wrapper.find(MemoryRouter).instance().history.location.search).toEqual('?request=111');
  });
});
