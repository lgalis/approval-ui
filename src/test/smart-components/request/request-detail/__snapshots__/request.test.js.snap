// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Request /> should render correctly 1`] = `
<Request
  indexpath={
    Object {
      "addComment": "/request/add-comment",
      "approve": "/request/approve",
      "deny": "/request/deny",
      "index": "/request",
    }
  }
  isExpanded={false}
  item={
    Object {
      "actions": Array [
        Object {
          "comments": null,
          "id": "1",
          "operation": "start",
          "processed_by": "system",
        },
        Object {
          "comments": null,
          "id": "2",
          "operation": "notify",
          "processed_by": "system",
        },
      ],
      "group_name": "Group1",
      "id": "111",
      "metadata": Object {
        "user_capabilities": Object {
          "approve": true,
          "deny": true,
        },
      },
      "state": "no-state",
    }
  }
  toggleExpand={[MockFunction]}
>
  <DataListItem
    aria-labelledby="check-request-111"
    isExpanded={false}
    key="request-111"
  >
    <li
      aria-labelledby="check-request-111"
      className="pf-c-data-list__item"
      id=""
    >
      <DataListItemRow
        key=".0"
        rowid="check-request-111"
      >
        <div
          className="pf-c-data-list__item-row"
        >
          <DataListToggle
            aria-label="Toggle details for"
            aria-labelledby="request-111 request-111"
            id="request-111"
            isExpanded={false}
            key=".0"
            onClick={[Function]}
            rowid="check-request-111"
          >
            <div
              aria-labelledby="request-111 request-111"
              className="pf-c-data-list__item-control"
              onClick={[Function]}
            >
              <div
                className="pf-c-data-list__toggle"
              >
                <Button
                  aria-controls={false}
                  aria-expanded={false}
                  aria-label="Toggle details for"
                  aria-labelledby={null}
                  id="request-111"
                  variant="plain"
                >
                  <button
                    aria-controls={false}
                    aria-disabled={false}
                    aria-expanded={false}
                    aria-label="Toggle details for"
                    aria-labelledby={null}
                    className="pf-c-button pf-m-plain"
                    data-ouia-component-id={0}
                    data-ouia-component-type="PF4/Button"
                    data-ouia-safe={true}
                    disabled={false}
                    id="request-111"
                    type="button"
                  >
                    <div
                      className="pf-c-data-list__toggle-icon"
                    >
                      <AngleRightIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                      >
                        <svg
                          aria-hidden={true}
                          aria-labelledby={null}
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style={
                            Object {
                              "verticalAlign": "-0.125em",
                            }
                          }
                          viewBox="0 0 256 512"
                          width="1em"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                            transform=""
                          />
                        </svg>
                      </AngleRightIcon>
                    </div>
                  </button>
                </Button>
              </div>
            </div>
          </DataListToggle>
          <DataListItemCells
            dataListCells={
              Array [
                <DataListCell>
                  <span
                    id="111-name"
                  >
                    Group1
                  </span>
                </DataListCell>,
                <DataListCell>
                  <span
                    id="111-state"
                    style={
                      Object {
                        "textTransform": "capitalize",
                      }
                    }
                  >
                    no-state
                  </span>
                </DataListCell>,
                <DataListCell>
                  <RequestActions
                    approveLink="/request/approve"
                    canApproveDeny={true}
                    canComment={true}
                    commentLink="/request/add-comment"
                    denyLink="/request/deny"
                    request={
                      Object {
                        "actions": Array [
                          Object {
                            "comments": null,
                            "id": "1",
                            "operation": "start",
                            "processed_by": "system",
                          },
                          Object {
                            "comments": null,
                            "id": "2",
                            "operation": "notify",
                            "processed_by": "system",
                          },
                        ],
                        "group_name": "Group1",
                        "id": "111",
                        "metadata": Object {
                          "user_capabilities": Object {
                            "approve": true,
                            "deny": true,
                          },
                        },
                        "state": "no-state",
                      }
                    }
                  />
                </DataListCell>,
              ]
            }
            key=".1"
            rowid="check-request-111"
          >
            <div
              className="pf-c-data-list__item-content"
            >
              <DataListCell
                key="111"
              >
                <div
                  className="pf-c-data-list__cell"
                >
                  <span
                    id="111-name"
                  >
                    Group1
                  </span>
                </div>
              </DataListCell>
              <DataListCell
                key="111-state"
              >
                <div
                  className="pf-c-data-list__cell"
                >
                  <span
                    id="111-state"
                    style={
                      Object {
                        "textTransform": "capitalize",
                      }
                    }
                  >
                    no-state
                  </span>
                </div>
              </DataListCell>
              <DataListCell
                key="111-action"
              >
                <div
                  className="pf-c-data-list__cell"
                >
                  <RequestActions
                    approveLink="/request/approve"
                    canApproveDeny={true}
                    canComment={true}
                    commentLink="/request/add-comment"
                    denyLink="/request/deny"
                    request={
                      Object {
                        "actions": Array [
                          Object {
                            "comments": null,
                            "id": "1",
                            "operation": "start",
                            "processed_by": "system",
                          },
                          Object {
                            "comments": null,
                            "id": "2",
                            "operation": "notify",
                            "processed_by": "system",
                          },
                        ],
                        "group_name": "Group1",
                        "id": "111",
                        "metadata": Object {
                          "user_capabilities": Object {
                            "approve": true,
                            "deny": true,
                          },
                        },
                        "state": "no-state",
                      }
                    }
                  >
                    <div
                      style={
                        Object {
                          "display": "flex",
                        }
                      }
                    />
                  </RequestActions>
                </div>
              </DataListCell>
            </div>
          </DataListItemCells>
        </div>
      </DataListItemRow>
      <DataListContent
        aria-label="Request Content Details"
        isHidden={true}
        key=".1"
        rowid="check-request-111"
      >
        <section
          aria-label="Request Content Details"
          className="pf-c-data-list__expandable-content"
          hidden={true}
          id=""
        >
          <div
            className="pf-c-data-list__expandable-content-body"
          >
            <Stack
              hasGutter={true}
            >
              <div
                className="pf-l-stack pf-m-gutter"
              >
                <StackItem>
                  <div
                    className="pf-l-stack__item"
                  >
                    <TextContent
                      component="h6"
                    >
                      <div
                        className="pf-c-content"
                        component="h6"
                      >
                        <ActionTranscript
                          actionList={
                            Array [
                              Object {
                                "comments": null,
                                "id": "1",
                                "operation": "start",
                                "processed_by": "system",
                              },
                              Object {
                                "comments": null,
                                "id": "2",
                                "operation": "notify",
                                "processed_by": "system",
                              },
                            ]
                          }
                        >
                          <Stack>
                            <div
                              className="pf-l-stack"
                            >
                              <div
                                key="1-action"
                              >
                                <TextContent>
                                  <div
                                    className="pf-c-content"
                                  >
                                    <Text
                                      className="pf-u-mb-0"
                                      component="small"
                                      key="1-action-created_at"
                                    >
                                      <small
                                        className="pf-u-mb-0"
                                        data-pf-content={true}
                                      >
                                        <span>
                                          <DateFormat
                                            type="relative"
                                          >
                                            Invalid date
                                          </DateFormat>
                                        </span>
                                      </small>
                                    </Text>
                                    <Text
                                      className="pf-u-mb-md"
                                      key="1-action-operation"
                                    >
                                      <p
                                        className="pf-u-mb-md"
                                        data-pf-content={true}
                                      >
                                        <OnRunningIcon
                                          color="currentColor"
                                          noVerticalAlign={false}
                                          size="sm"
                                        >
                                          <svg
                                            aria-hidden={true}
                                            aria-labelledby={null}
                                            fill="currentColor"
                                            height="1em"
                                            role="img"
                                            style={
                                              Object {
                                                "verticalAlign": "-0.125em",
                                              }
                                            }
                                            viewBox="0 64 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M511.781-73.143c-70.656 0-136.923 13.312-198.875 40.009s-116.078 63.269-162.45 109.934c-46.299 46.665-82.944 100.937-109.934 162.962s-40.448 128.293-40.448 198.875 13.458 136.923 40.448 198.875 63.634 116.078 109.934 162.45c46.299 46.299 100.425 82.944 162.45 109.934s128.293 40.448 198.875 40.448 136.923-13.458 198.875-40.448 116.078-63.634 162.45-109.934c46.299-46.299 82.944-100.425 109.934-162.45s40.448-128.293 40.448-198.875-13.458-136.923-40.448-198.875-63.634-116.297-109.934-162.962-100.425-83.31-162.45-109.934-128.293-40.009-198.875-40.009v0zM361.326 795.94c-46.958-20.334-88.137-48.128-123.465-83.456s-63.122-76.434-83.456-123.465-30.501-97.134-30.501-150.455 10.167-103.643 30.501-150.894c20.334-47.323 48.128-88.43 83.456-123.465s76.434-62.61 123.465-82.944 97.134-30.501 150.455-30.501 103.424 10.167 150.455 30.501c46.958 20.334 88.137 47.982 123.465 82.944s63.122 76.142 83.456 123.465 30.501 97.646 30.501 150.894-10.167 103.424-30.501 150.455c-20.334 46.958-48.128 88.137-83.456 123.465s-76.434 63.122-123.465 83.456-97.134 30.501-150.455 30.501c-53.321 0-103.424-10.167-150.455-30.501v0zM730.697 498.395c1.975-8.119 2.414-19.383-1.17-26.331-7.387-14.263-16.457-26.99-26.185-40.96-83.017-119.296-158.135-200.485-249.71-305.371-17.993-20.626-34.889-13.751-42.057-9.070-15.579 10.24-15.799 29.769-7.314 52.297 9.801 25.966 20.48 57.783 29.842 79.799 22.455 52.736 20.187 58.075 41.911 114.907-52.297-0.1024-127.781-0.951-180.078-1.463-15.067-0.146-33.499 1.902-39.497 16.311-5.778 13.751-0.146 26.039 8.558 38.107 64.146 89.381 172.617 212.626 248.174 294.546 6.51 7.022 30.062 23.333 43.739 21.358 10.825-1.536 21.87 1.243 26.843-9.289 5.851-12.434 2.706-25.746-2.56-37.23-22.016-48.347-45.714-118.638-71.095-174.519 50.615-0.073 144.53-0.073 195.145-0.146 8.558 0 18.432-0.731 23.406-8.046 1.024-1.536 1.755-3.145 2.121-4.901v0z"
                                              transform="rotate(180 0 512) scale(-1 1)"
                                            />
                                          </svg>
                                        </OnRunningIcon>
                                         
                                        Started by system
                                      </p>
                                    </Text>
                                  </div>
                                </TextContent>
                              </div>
                              <div
                                key="2-action"
                              >
                                <TextContent>
                                  <div
                                    className="pf-c-content"
                                  >
                                    <Text
                                      className="pf-u-mb-0"
                                      component="small"
                                      key="2-action-created_at"
                                    >
                                      <small
                                        className="pf-u-mb-0"
                                        data-pf-content={true}
                                      >
                                        <span>
                                          <DateFormat
                                            type="relative"
                                          >
                                            Invalid date
                                          </DateFormat>
                                        </span>
                                      </small>
                                    </Text>
                                    <Text
                                      className="pf-u-mb-md"
                                      key="2-action-operation"
                                    >
                                      <p
                                        className="pf-u-mb-md"
                                        data-pf-content={true}
                                      >
                                        <EnvelopeIcon
                                          color="currentColor"
                                          noVerticalAlign={false}
                                          size="sm"
                                        >
                                          <svg
                                            aria-hidden={true}
                                            aria-labelledby={null}
                                            fill="currentColor"
                                            height="1em"
                                            role="img"
                                            style={
                                              Object {
                                                "verticalAlign": "-0.125em",
                                              }
                                            }
                                            viewBox="0 0 512 512"
                                            width="1em"
                                          >
                                            <path
                                              d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
                                              transform=""
                                            />
                                          </svg>
                                        </EnvelopeIcon>
                                         
                                        Notified by system
                                      </p>
                                    </Text>
                                  </div>
                                </TextContent>
                              </div>
                            </div>
                          </Stack>
                        </ActionTranscript>
                      </div>
                    </TextContent>
                  </div>
                </StackItem>
              </div>
            </Stack>
          </div>
        </section>
      </DataListContent>
    </li>
  </DataListItem>
</Request>
`;
